### PR TITLE
[WIP] [Feature] Support _state/_merge/_union combinator

### LINKED ISSUE #49000: Support Generic Aggregate Function State
- [x] https://github.com/StarRocks/starrocks/pull/50069
- [x] https://github.com/StarRocks/starrocks/pull/50425
- [x] https://github.com/StarRocks/starrocks/pull/51510

## Feature request

**Is your feature request related to a problem? Please describe.**

When the detail data is very large, the aggregation model is very useful for querying after summarizing by dimensions; moreover, the aggregation model is also the necessary means for incremental computation of aggregate functions in StarRocks at present. 

However, it currently only supports the following aggregate functions, which are very limited; even in terms of extensibility, if you want to support intermediate states such as HllSketch or Avg functions, the current development support logic is also quite complicated, with a lot of repetitive work.

**Describe the solution you'd like**

Since all aggregate functions have already defined the types of their intermediate states and methods for serialization/deserialization, all aggregate functions defined in the Query Engine can be used as columns in the aggregation model, and are no longer limited to the aggregate functions currently supported by the aggregation model.

#### Support Generic Aggregate Function in Aggregate Model
- https://github.com/StarRocks/starrocks/pull/50069

```
ColName aggregateFunctionName(paramType1, paramType2, ....)
```

- ColName: The column name when the column is stored.
- aggregateFunctionName: Defines the aggregate function corresponding to this column, used for compaction and re-aggregation.
- The type of this column is the state type of the aggregate function, which will be automatically inferred.
Since obtaining an aggregate function requires the function's input parameters + return type + whether it is nullable, the input parameters of the aggregate function need to be input here to uniquely determine an aggregate function.

#### Support Aggregate Function Combinator
- https://github.com/StarRocks/starrocks/pull/50425

- Use combinator to support common aggregate functions transforms:

- `_state` combinator
```
/**
 * DESC: immediate_type {agg_func}_state(arg_types)
 *  input type  : aggregate function's argument types
 *  return type : aggregate function's immediate type
 */
```

- `_union` combinator
```

/**
 * Union combinator for aggregate function to union the agg state to return the immediate result of aggregate function.
 * DESC: immediate_type {agg_func}_union(immediate_type)
 *  input type          : aggregate function's immediate_type
 *  intermediate type   : aggregate function's immediate_type
 *  return type         : aggregate function's immediate_type
 */
```

- `_merge` combinator
```
/**
 * Merge combinator for aggregate function to merge the agg state to return the final result of aggregate function.
 * DESC: return_type {agg_func}_merge(immediate_type)
 *  input type          : aggregate function's immediate_type
 *  intermediate type   : aggregate function's immediate_type
 *  return type         : aggregate function's return type
 */
```
#### Support to Use Generic Aggregate Function State in Synchronized Materialized View
- Support create & rewrite all common aggregate functions in synchronized materialized view

#### Support to Use Generic Aggregate Function State in ASynchronized Materialized View
- Support create & rewrite all common aggregate functions in asynchronized materialized view

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by LiShuMing**
## What I'm doing:

Fixes https://github.com/StarRocks/starrocks/issues/49000

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [x] Yes, this PR will result in a change in behavior.
- [ ] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 [hunk 1]
```diff
@@ -2615,6 +2625,7 @@ type
     | arrayType
     | structType
     | mapType
+    | aggStateType
```
Comment (alvin-celerdata) [inline fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:2628]: I think we need to discuss the type first.
I don't think there is a need to introduce a new type, we can use binary here.

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 [hunk 2]
```diff
@@ -425,6 +434,7 @@ aggDesc
     | BITMAP_UNION
     | PERCENTILE_UNION
     | REPLACE_IF_NOT_NULL
+    | AGG_STATE_UNION
```
Comment (alvin-celerdata) [inline fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:437]: We can use an aggregation function here.

**Review by alvin-phoenix-ai — CHANGES_REQUESTED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 (hunk 1 above)
Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:2628]: This pr doesn't add a new type instead.

You can see it deeply, this is only a syntax sugar which wraps an aggregate function's intermediate type which is not seen by users.

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 (hunk 2 above)
Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:437]: Yes, `AGG_STATE_UNION` is a generic aggregate function which wraps all current aggregate functions which its input is an intermediate type.

**Pull request closed without merging**